File: code/ex956.cpp
class Solution {
public:
    int tallestBillboard(vector<int>& rods) {
        int maxdiff = 0;
        for (auto& rod : rods)    maxdiff += rod;
        auto memo = vector<vector<int>>(maxdiff + 1, vector<int>(rods.size()));
        return recurrsion(rods, 0, 0, memo);
    }
    
    int recurrsion(vector<int>& rods, int diff, int index, vector<vector<int>>& memo) {
        if (index >= rods.size()) {
            if (diff != 0)    return INT_MIN;
            else    return 0;
        }
        
        // Load memo
        if (memo[diff][index] != 0)    return memo[diff][index];
        
        int addlongleg = recurrsion(rods, diff + rods[index], index + 1, memo);
        int addshortleg = recurrsion(rods, std::abs(diff - rods[index]), index + 1, memo) + min(diff, rods[index]);
        int notadd = recurrsion(rods, diff, index + 1, memo);
        
        // Save memo
        memo[diff][index] = max(addlongleg, max(addshortleg, notadd));
        return memo[diff][index];
    }
};

File: code/ex956.java
// Backtracking with memo

class Solution {
    public int n;
    
    public int tallestBillboard(int[] rods) {
        n = rods.length;
        
        int sum = 0;
        for (int rod : rods)    sum += rod;
        
        int[][] memo = new int[n][sum + 1];
        return maxHeight(rods, 0, 0, memo);
    }
    
    private int maxHeight(int[] rods, int s, int diff, int[][] memo) {
        if (s >= n)    return diff == 0 ? 0 : Integer.MIN_VALUE;
        
        if (memo[s][diff] != 0)    return memo[s][diff];
        // Three scenarios: Add to the longer leg, add to the shorter leg, do not add to any leg
        int addToLong = maxHeight(rods, s + 1, diff + rods[s], memo);
        int addToShort = maxHeight(rods, s + 1, Math.abs(diff - rods[s]), memo) + Math.min(rods[s], diff);    // Add the length of part with same length to the result
        int notAdd = maxHeight(rods, s + 1, diff, memo);
        
        int longest = Math.max(Math.max(addToLong, addToShort), notAdd);
        memo[s][diff] = longest;
        return longest;
    }
}